light-client-poc/src/circuit/state_update.rs

```diff
@@ -0,0 +1,665 @@
+#![allow(unused_imports)]
+
+use eth_types::Field;
+use eyre::Result;
+use gadgets::{
+    is_zero::{IsZeroChip, IsZeroConfig, IsZeroInstruction},
+    util::{
+        and,
+        not::{self},
+        or, Expr,
+    },
+};
+use halo2_proofs::{
+    circuit::{Layouter, SimpleFloorPlanner, Value},
+    dev::MockProver,
+    halo2curves::bn256::Fr,
+    plonk::{
+        Advice, Circuit, Column, ConstraintSystem, Error, Expression, Fixed, Instance, ProvingKey,
+        Selector,
+    },
+    poly::Rotation,
+    SerdeFormat,
+};
+use rand::SeedableRng;
+use rand_chacha::ChaCha20Rng;
+use std::time::Instant;
+use zkevm_circuits::{
+    keccak_circuit::{KeccakCircuit, KeccakCircuitConfig, KeccakCircuitConfigArgs},
+    mpt_circuit::{MPTCircuit, MPTCircuitParams, MPTConfig},
+    table::{KeccakTable, MptTable},
+    util::{word, Challenges, SubCircuit, SubCircuitConfig},
+};
+
+pub const DEFAULT_MAX_PROOF_COUNT: usize = 20;
+pub const DEFAULT_CIRCUIT_DEGREE: usize = 14;
+
+use super::witness::{
+    PublicInputs, SingleTrieModification, SingleTrieModifications, StateUpdateWitness, Transforms,
+};
+
+use halo2_proofs::{
+    halo2curves::bn256::{Bn256, G1Affine},
+    plonk::{create_proof, keygen_pk, keygen_vk, verify_proof},
+    poly::{
+        commitment::ParamsProver,
+        kzg::{
+            commitment::{KZGCommitmentScheme, ParamsKZG, ParamsVerifierKZG},
+            multiopen::{ProverSHPLONK, VerifierSHPLONK},
+            strategy::SingleStrategy,
+        },
+    },
+    transcript::{
+        Blake2bRead, Blake2bWrite, Challenge255, TranscriptReadBuffer, TranscriptWriterBuffer,
+    },
+};
+
+// A=>B  eq ~(A & ~B) (it is not the case that A is true and B is false)
```

Comment:
I am a bit confused - the comment says it correctly, but then the implementation seems to be `A & ~B` instead of `~(A & ~B)`?

Reply:
Ah, yes! :)